Microsoft.Research/AbstractInterpretation/Abstract Domains/Array/ArrayState.cs

```diff
@@ -22,388 +10,388 @@
+        }
+
+        public ArrayAbstractDomain Array
+        {
+            get
+            {
+                Contract.Ensures(Contract.Result<ArrayAbstractDomain>() != null);
+
+                return this.Right;
+            }
+        }
+
+        public object OtherAbstractStates
+        {
+            get
+            {
+                object value;
+                _additionalState.HasValue(out value);
+
+                return value;
+            }
+        }
+
+        public ArrayStateAD<NumericalAbstractDomain, ArrayAbstractDomain, Variable, Expression>
+          MakeNewWithUpdatedArrayState(ArrayAbstractDomain newState)
+        {
+            Contract.Requires(newState != null);
+
+            return new ArrayStateAD<NumericalAbstractDomain, ArrayAbstractDomain, Variable, Expression>
+              (this.Numerical, newState, _expManager);
+        }
+
+
+        #region Implementation of abstract methods
+
+        public override ReducedCartesianAbstractDomain<NumericalAbstractDomain, ArrayAbstractDomain> Reduce(NumericalAbstractDomain left, ArrayAbstractDomain right)
+        {
+            return this.Factory(left, right);
+        }
+
+        protected override ReducedCartesianAbstractDomain<NumericalAbstractDomain, ArrayAbstractDomain> Factory(NumericalAbstractDomain left, ArrayAbstractDomain right)
+        {
+            Contract.Ensures(Contract.Result<ReducedCartesianAbstractDomain<NumericalAbstractDomain, ArrayAbstractDomain>>() != null);
+
+            return new ArrayStateAD<NumericalAbstractDomain, ArrayAbstractDomain, Variable, Expression>(left, right, _expManager);
+        }
 
-    #endregion
+        static public ArrayStateAD<NumericalAbstractDomain, ArrayAbstractDomain, Variable, Expression> For(
+          NumericalAbstractDomain left, ArrayAbstractDomain right,
+           ExpressionManager<Variable, Expression> expManager)
+        {
+            Contract.Requires(left != null);
+            Contract.Requires(right != null);
+            Contract.Requires(expManager != null);
 
-    #region IAbstractDomain Members
+            Contract.Ensures(Contract.Result<ArrayStateAD<NumericalAbstractDomain, ArrayAbstractDomain, Variable, Expression>>() != null);
 
-    bool IAbstractDomain.LessEqual(IAbstractDomain a)
-    {
-      return base.LessEqual(a);
-    }
+            return new ArrayStateAD<NumericalAbstractDomain, ArrayAbstractDomain, Variable, Expression>(left, right, expManager);
+        }
 
-    public bool LessEqual(ArrayStateAD<NumericalAbstractDomain, ArrayAbstractDomain, Variable, Expression> a)
-    {
-      Contract.Requires(a  != null);
+        [SuppressMessage("Microsoft.Contracts", "RequiresAtCall-!(other is Optional)", Justification = "We do not track dynamic types")]
+        static public ArrayStateAD<NumericalAbstractDomain, ArrayAbstractDomain, Variable, Expression> For<OtherAbstractDomain>(
+          NumericalAbstractDomain left, ArrayAbstractDomain right, OtherAbstractDomain other,
```

Comment:
This is part of a continuation indent that Visual Studio will ignore. It's not "absolutely perfect", but it won't cause problems either.